### ISSUE TITLE
Using opm-grid as dune-module fails during CMake

### ISSUE THREAD

**Issue by blattms** [OPM/opm-grid#598]
This one of the automated tests for Debian and it fails with:
```
-- Performing Test HAVE_DUNE_GEOMETRY
CMake Error at /tmp/autopkgtest.WoaAQI/autopkgtest_tmp/dune-autopkgtest/build-cmake/CMakeFiles/CMakeTmp/cmTC_4d2e6Targets.cmake:21 (set_target_properties):
  The link interface of target "dunecommon" contains:

    Threads::Threads

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /tmp/autopkgtest.WoaAQI/autopkgtest_tmp/dune-autopkgtest/build-cmake/CMakeFiles/CMakeTmp/CMakeLists.txt:16 (include)


CMake Error at /tmp/autopkgtest.WoaAQI/autopkgtest_tmp/dune-autopkgtest/build-cmake/CMakeFiles/CMakeTmp/cmTC_4d2e6Targets.cmake:32 (set_target_properties):
  The link interface of target "dunegeometry" contains:

    Threads::Threads

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /tmp/autopkgtest.WoaAQI/autopkgtest_tmp/dune-autopkgtest/build-cmake/CMakeFiles/CMakeTmp/CMakeLists.txt:16 (include)


CMake Error at /usr/share/cmake-3.24/Modules/Internal/CheckSourceCompiles.cmake:95 (try_compile):
  Failed to generate test project build system.
Call Stack (most recent call first):
  /usr/share/cmake-3.24/Modules/CheckCXXSourceCompiles.cmake:76 (cmake_check_source_compiles)
  /usr/share/opm/cmake/Modules/OpmPackage.cmake:139 (check_cxx_source_compiles)
  /usr/share/opm/cmake/Modules/Finddune-geometry.cmake:14 (find_opm_package)
  /usr/share/opm/cmake/Modules/OpmFind.cmake:142 (find_package)
  /usr/share/opm/cmake/Modules/OpmPackage.cmake:192 (find_and_append_package_to)
  /usr/share/opm/cmake/Modules/OpmPackage.cmake:87 (find_package_deps)
  /usr/share/opm/cmake/Modules/Finddune-grid.cmake:17 (find_opm_package)
  /usr/share/opm/cmake/Modules/OpmFind.cmake:142 (find_package)
  /usr/share/opm/cmake/Modules/OpmPackage.cmake:192 (find_and_append_package_to)
  /usr/share/opm/cmake/Modules/opm-grid-prereqs.cmake:32 (find_package_deps)
  /usr/lib/x86_64-linux-gnu/cmake/opm-grid/opm-grid-config.cmake:119 (include)
  /usr/share/dune/cmake/modules/DuneMacros.cmake:203 (find_package)
  /usr/share/dune/cmake/modules/DuneMacros.cmake:451 (find_dune_package)
  /usr/share/dune/cmake/modules/DuneMacros.cmake:494 (dune_process_dependency_leafs)
  /usr/share/dune/cmake/modules/DuneMacros.cmake:518 (dune_create_dependency_leafs)
  /usr/share/dune/cmake/modules/DuneMacros.cmake:666 (dune_create_dependency_tree)
  CMakeLists.txt:19 (dune_project)


-- Configuring incomplete, errors occurred!
See also "/tmp/autopkgtest.WoaAQI/autopkgtest_tmp/dune-autopkgtest/build-cmake/CMakeFiles/CMakeOutput.log".
--- Failed to build dune-autopkgtest ---
```
Marking this as 2022.10 as it is not Debian specific. Note that building the module succeeds, though.

**Comment by blattms** [comment]
Bugger, this is even more broken with DUNE 2.7:
+ Somehow a colon sneak in into the library name
```
opm-grid_STATIC_LIBRARIES:INTERNAL=pthread;:dunegrid;:dunealbertagrid_1d;:dunealbertagrid_2d;:dunealbertagrid_3d;:dunegeometry;:ugL;:ugS2;:ugS3;alberta_1d;alberta_2d;alberta_3d;alberta_utilities;ltdl;:dunecommon;mpi_cxx;superlu;umfpack;amd;lapack;blas;f77blas;atlas;m;suitesparseconfig;rt;:libopmcommon.a;python3.9;gomp;pthread;:Boost::headers;cjson;trilinos_zoltan;parmetis;metis;mpi;scotch;ptscotch;ptscotcherr;:libopmgrid.a
```

For all versions if our modules are not installed than the CMAKE_MODULES_PATH in opm-module-config.cmake is incorrect (pointing to the local source directory instead of that for opm-common) and we will not find OpmPackage.cmeke

**Comment by blattms** [comment]
Ok, ir works if I have all dependencies installed, set CMAKE_PREFIX_PATH to include their locations and use plain cmake. Other scenarios are quite broken.

**Comment by atgeirr** [comment]
I do not really understand what is wrong here. Do you have an idea how to resolve this? If not, then we should probably not let it hold the release.

**Comment by blattms** [comment]
Well, the dependency chain seems to be broken. Dependencies are searched after the module itself, it seems. Probably only an issue for DUNE modules using OPM modules.

**Comment by blattms** [comment]
It is  a mess. 
What works for DUNE 2.7 won't work for DUNE 2.8/2.9 and vice versa. I think I could hack around it by reverting the fix of the if-statement in OPM/opm-common@20e3202067 Before that compilation was never tested and all was fine :rofl: 

Now if compilation is tested with DUNE >=2.8 the imported targets of the dependencies have to be there, but they are not if we don't use the full OPM build system. That is because first we find the a package that we depend on, and then we search for the dependencies it needs.

Turning that order around breaks with DUNE 2.7 because then the dependencies are not included in module_{LIBRARIES|INCLUDE_DIRS}. E.g. dune-istl_INCLUDE_DIRS will be missing the include directories of suitesparse..

That means I know have a fix for Debian unstable but that will break Ubuntu :see_no_evil: , I guess I will have to activate the compilation test again. :grimacing:

**Comment by blattms** [comment]
> It is a mess. What works for DUNE 2.7 won't work for DUNE 2.8/2.9 and vice versa. I think I could hack around it by reverting the fix of the if-statement in [OPM/opm-common@20e3202067](https://github.com/OPM/opm-common/commit/20e3202067) Before that compilation was never tested and all was fine rofl
> 

I verified that said hack makes the test go through.
